### DISCUSSION TITLE
Internal node packages? (NestJS)

### DISCUSSION THREAD

**Discussion by Scalahansolo** [vercel/turborepo#5341] [Help] [answered]
The docs over at https://turbo.build/repo/docs/handbook/sharing-code/internal-packages are great. I was hoping to build out a repo using internal packages for my NestJS apps. I was curious if it would be possible to update the instructions at https://turbo.build/repo/docs/handbook/sharing-code/internal-packages#6-configuring-your-app to show using something like webpack to work with internal packages?

**Comment by mohitxskull** [comment]
Any update on this?

i have a honojs app and i have a package which is exporting schemas and types 

its working fine nextjs ( becuase of transpile option )

**Comment by anthonyshew** [comment, marked as answer]
We've updated our guidance on creating Internal Packages. Thinking it will help on this: https://turbo.build/repo/docs/crafting-your-repository/creating-an-internal-package

**Reply by StringKe** [reply]
The current answer is not a good solution for CJS and ESM in internal packages.

First of all my internal package cannot be distributed, assuming you have a `@repo/utils` that can be used by both react and nestjs.

1. if I compile the package, then the nestjs compilation won't include it, it won't exist at runtime.
2. if I use the `Just-in-Time Packages` package in the docs, it will be considered as an ESM and the CJS that nestjs requires will conflict with the typescript configuration.

There needs to be a way for nestjs, or any other webpack or vite or whatever, to include internal (non-published) packages in the build.

docs1: https://turbo.build/repo/docs/core-concepts/internal-packages#compilation-strategies
docs2: https://turbo.build/repo/docs/crafting-your-repository/creating-an-internal-package

**Reply by anthonyshew** [reply]
> There needs to be a way for nestjs, or any other webpack or vite or whatever, to include internal (non-published) packages in the build.

The documentation describes exactly this. Can you be more specific about the problem you're having?

**Reply by StringKe** [reply]
I've tried a lot of methods that have solved my needs.

Current packaging tools do not support automatic handling of these internal packages for the time being (the Just-in-Time Packages approach to ESM and CJS coexistence), and should be documented to indicate that additional steps need to be configured.

Finally I need to configure the way all my apps are packaged

```js
const nodeExternals = require('webpack-node-externals');

module.exports = (options, webpack) => {
  return {
    ...options,
    externals: [
      nodeExternals({
        allowlist: [/^@app/],
      }),
    ],
    devtool: 'source-map',
  };
};
```

My Solution.

1. all packages under packages are compiled with tsup for esm and cjs.
2. apps are forced to package dependencies starting with @app by configuring the compilation tool.

**Reply by anthonyshew** [reply]
We do cover the general approach, but some tools may require additional steps or have their intricacies.

If you'd like to help us have information in the docs for your edge case, that'd be great! Happy to review your PR!
